distribution_metadata.c

```diff
@@ -524,62 +467,68 @@ ColumnNameToColumn(Oid relationId, char *columnName)
 
 
 /*
- * LoadShardIntervalRow finds the row for the specified shard identifier in the
- * shard table and copies values from that row into the provided output params.
+ * TupleToShardPlacement populates a ShardInterval using values from a row of
+ * the shard configuration table and returns a pointer to that struct. The
+ * input tuple must not contain any NULLs.
+ *
+ * Because the partition type and column must be known in order to interpret
+ * the min/max values of a shard interval, the provided HeapTuple must contain
+ * these fields. This is easily accomplished by JOINing to the partition table.
  */
-static void
-LoadShardIntervalRow(int64 shardId, Oid *relationId, char **minValue,
-					 char **maxValue)
+static ShardInterval *
+TupleToShardInterval(HeapTuple heapTuple, TupleDesc tupleDescriptor)
 {
-	RangeVar *heapRangeVar = NULL;
-	RangeVar *indexRangeVar = NULL;
-	Relation heapRelation = NULL;
-	Relation indexRelation = NULL;
-	IndexScanDesc indexScanDesc = NULL;
-	const int scanKeyCount = 1;
-	ScanKeyData scanKey[scanKeyCount];
-	HeapTuple heapTuple = NULL;
-
-	heapRangeVar = makeRangeVar(METADATA_SCHEMA_NAME, SHARD_TABLE_NAME, -1);
-	indexRangeVar = makeRangeVar(METADATA_SCHEMA_NAME, SHARD_PKEY_INDEX_NAME, -1);
-
-	heapRelation = relation_openrv(heapRangeVar, AccessShareLock);
-	indexRelation = relation_openrv(indexRangeVar, AccessShareLock);
+	ShardInterval *shardInterval = NULL;
+	bool isNull = false;
+	Oid intervalTypeId = InvalidOid;
+	int32 intervalTypeMod = -1;
+	Oid inputFunctionId = InvalidOid;
+	Oid typeIoParam = InvalidOid;
 
-	ScanKeyInit(&scanKey[0], 1, BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(shardId));
+	Datum idDatum = SPI_getbinval(heapTuple, tupleDescriptor, 1, &isNull);
```

Comment:
> It doesn't need to be recursive right? Could we not have the callers call PartitionColumn, and pass it as an argument to this function? I feel like that gives us a better separation, rather than the query joining with the partition table. The caller then explicitly knows the function needs a partition column.

Sorry, by recursive, I meant calling an SPI-using function from within another SPI-enabled section. But what you're describing sounds great. I'll do that.